src/shell/shell_cmds.cpp

```diff
@@ -494,236 +725,166 @@ void DOS_Shell::CMD_DIR(char * args) {
 		args=const_cast<char*>(line.c_str());
 	}
 
+	ScanCMDBool(args,"4"); /* /4 ignored (default) */
 	bool optW=ScanCMDBool(args,"W");
-	ScanCMDBool(args,"S");
 	bool optP=ScanCMDBool(args,"P");
 	if (ScanCMDBool(args,"WP") || ScanCMDBool(args,"PW")) {
 		optW=optP=true;
 	}
+	if (ScanCMDBool(args,"-W")) optW=false;
+	if (ScanCMDBool(args,"-P")) optP=false;
+	bool optZ=ScanCMDBool(args,"Z");
+	if (ScanCMDBool(args,"-Z")) optZ=false;
+	bool optS=ScanCMDBool(args,"S");
+	if (ScanCMDBool(args,"-S")) optS=false;
 	bool optB=ScanCMDBool(args,"B");
-	bool optAD=ScanCMDBool(args,"AD");
+	if (ScanCMDBool(args,"-B")) optB=false;
+	bool optA=ScanCMDBool(args,"A");
+	bool optAD=ScanCMDBool(args,"AD")||ScanCMDBool(args,"A:D");
 	bool optAminusD=ScanCMDBool(args,"A-D");
+	bool optAS=ScanCMDBool(args,"AS")||ScanCMDBool(args,"A:S");
+	bool optAminusS=ScanCMDBool(args,"A-S");
+	bool optAH=ScanCMDBool(args,"AH")||ScanCMDBool(args,"A:H");
+	bool optAminusH=ScanCMDBool(args,"A-H");
+	bool optAR=ScanCMDBool(args,"AR")||ScanCMDBool(args,"A:R");
+	bool optAminusR=ScanCMDBool(args,"A-R");
+	bool optAA=ScanCMDBool(args,"AA")||ScanCMDBool(args,"A:A");
+	bool optAminusA=ScanCMDBool(args,"A-A");
+	if (ScanCMDBool(args,"-A")) {
+		optA = false;
+		optAD = false;
+		optAminusD = false;
+		optAS = false;
+		optAminusS = false;
+		optAH = false;
+		optAminusH = false;
+		optAR = false;
+		optAminusR = false;
+		optAA = false;
+		optAminusA = false;
```

Comment:
There's a lot of mapping going on here. I wonder something like a `std::map` between these mode and attributes, mapping over to bools or strings, might let you boil all of this down to a single call. 

Ie: you'd have `std::map<LFNAttrs, bool> lfn_option_map` with enum keys: AS, AH, AR, AA, AminuS, and so on.

Reply:
@kcgen These are not really LFN(-only) attributes, but attributes for both SFN and LFN. They are the actual options for DIR command as in MS-DOS. I largely followed the existing format for handling options.

Reply:
Ok, understood.. I'm just suggesting that a map could significantly clean this up, regardless how it was originally coded or if LFN contributed a little or a lot to this pile of logic.

Perhaps this would be too risky for the current PR and would go into a cleanup PR; just an option.

Reply:
I think a cleanup PR will be desired.